### PR TITLE
udevd: implement dobule-forking in daemon mode

### PR DESCRIPTION
**Pull request by teg**
As long as we support non-systemd systems, we likely need to support the
legacy --daemon mode, though the intention is to drop it eventually.

This patch improves the current (single-forking) implementation in three ways:
- it fully implements the recomendation from daemon(7)
- it has exactly the same notion of readiness in daemon and notify mode
- it keeps all the forking logic in one function rather than mixing it up with
  the non-legacy logic in main().

This fixes issue #190 (by essentially reverting 5c67cf2).

### CODE REVIEW

**Comment by poettering** [comment]
Hmm, most distros used to have a tool like start-stop-daemon, that can be used for this. Maybe we should strip all daemonization code and simply ask downstream distros to use that if they don't have an init system like systemd or upstart that can handle this for them?

**Comment by teg** [comment]
If there was a tool that could handle this for us, that would be great. But I guess it would then need to be able to integrate with sd_notify() to get the synchronization correctly, which I guess third-party tools don't do. Suggestions welcome.

File: src/udev/udevd.c [hunk 1]
```diff
@@ -1735,6 +1691,147 @@ int main(int argc, char *argv[]) {
         if (manager)
                 udev_ctrl_cleanup(manager->ctrl);
         mac_selinux_finish();
+
+        return r;
+}
+
+static int main_forking(void) {
+        Barrier barrier = {};
+        _cleanup_(barrier_destroyp) Barrier *b = &barrier;
```
Comment (dvdhrm) [inline src/udev/udevd.c:1700]: What's that? Why not replace both lines with:
`_cleanup_(barrier_destroy) Barrier barrier = BARRIER_NULL;`

No reason to use 'barrier_destroyp'.

File: src/udev/udevd.c [hunk 2]
```diff
@@ -1735,6 +1691,147 @@ int main(int argc, char *argv[]) {
...
+                        /* child */
+                        log_close();
+                        _exit(EXIT_SUCCESS);
+                }
+                break;
+        default:
+                /* parent */
+                barrier_set_role(&barrier, BARRIER_PARENT);
+
+                r = barrier_wait_next(&barrier);
```
Comment (dvdhrm) [inline src/udev/udevd.c:1784]: This should be:
`r = barrier_place_and_sync(&barrier);`

I should really improve the documentation and drop all the `barrier_*_next()` helpers.. they should really not be used by anyone.

File: src/udev/udevd.c (hunk 1 above)
Comment (teg) [inline src/udev/udevd.c:1700]: Indeed, I got the DEFINE_TRIVIAL_FUNC wrong.

**Comment by teg** [comment]
This was likely overkill, I'll propose a simpler solution.

**Pull request closed without merging**